service/history/workflow/mutable_state_impl.go

```diff
@@ -6252,3 +6259,259 @@ func (ms *MutableStateImpl) RefreshExpirationTimeoutTask(ctx context.Context) er
+			if err != nil {
+				// we don't have enough information to recreate all parents
+				return err
+			}
+			currentNode, err = parentNode.AddChild(incomingPath[len(incomingPath)-1], nodeMutation.Data)
+			if err != nil {
+				return err
+			}
+		}
+		internalNode := currentNode.InternalRepr()
+		if CompareVersionedTransition(nodeMutation.LastUpdateVersionedTransition, internalNode.LastUpdateVersionedTransition) <= 0 {
+			continue
+		}
+		internalNode.Data = nodeMutation.Data
+		internalNode.InitialVersionedTransition = nodeMutation.InitialVersionedTransition
+		internalNode.LastUpdateVersionedTransition = nodeMutation.LastUpdateVersionedTransition
+		internalNode.TransitionCount++
+	}
+	ms.mustInitHSM()
+	return nil
+}
+
+func (ms *MutableStateImpl) applySignalRequestedIds(signalRequestedIds []string, incomingExecutionInfo *persistencespb.WorkflowExecutionInfo) {
+	incomingVersionedTransition := incomingExecutionInfo.TransitionHistory[len(incomingExecutionInfo.TransitionHistory)-1]
+	if CompareVersionedTransition(incomingVersionedTransition, ms.executionInfo.SignalRequestIdsLastUpdateVersionedTransition) > 0 {
+		for requestID := range ms.pendingSignalRequestedIDs {
+			ms.approximateSize -= len(requestID)
+		}
+		ms.pendingSignalRequestedIDs = convert.StringSliceToSet(signalRequestedIds)
+		for requestID := range ms.pendingSignalRequestedIDs {
+			ms.approximateSize += len(requestID)
+		}
+		ms.executionInfo.SignalRequestIdsLastUpdateVersionedTransition = incomingVersionedTransition
+	}
+}
+
+func applyUpdatesToSubStateMachine[K comparable, V lastUpdatedStateTransitionGetter](
+	ms *MutableStateImpl,
+	pendingInfos map[K]V,
+	updateInfos map[K]V,
+	updatedSubStateMachine map[K]V,
+	versionedTransition *persistencespb.VersionedTransition,
+	fn func(current, incoming V),
+) {
+	for key, updated := range updatedSubStateMachine {
+		if CompareVersionedTransition(updated.GetLastUpdateVersionedTransition(), versionedTransition) <= 0 {
+			continue
+		}
+		val := updated
+		var existing V
+		if existing, ok := pendingInfos[key]; ok {
+			ms.approximateSize -= existing.Size() + int64SizeBytes
+		}
+		if fn != nil {
+			val = common.CloneProto(updated)
+			fn(existing, val)
+		}
+		pendingInfos[key] = val
+		updateInfos[key] = val
+		ms.approximateSize += val.Size() + int64SizeBytes
+	}
+}
+
+func (ms *MutableStateImpl) applyUpdatesToUpdateInfos(
+	updatedUpdateInfos map[string]*persistencespb.UpdateInfo,
+	versionedTransition *persistencespb.VersionedTransition,
+) {
+	for updateID, ui := range updatedUpdateInfos {
+		if CompareVersionedTransition(ui.GetLastUpdateVersionedTransition(), versionedTransition) <= 0 {
+			continue
+		}
+		if existing, ok := ms.executionInfo.UpdateInfos[updateID]; ok {
+			ms.approximateSize -= existing.Size() + len(updateID)
+		} else {
+			ms.executionInfo.UpdateCount++
+		}
+		ms.executionInfo.UpdateInfos[updateID] = ui
+		ms.approximateSize += ui.Size() + len(updateID)
+		ms.updateInfoUpdated[updateID] = struct{}{}
```

Comment:
nit: this field doesn't have to be updated, it's not used for constructing mutable state mutation when closing transaction.